### PR TITLE
GCP provisioning: handle 'wait_ready timeout exceeded'.

### PR DESCRIPTION
**Pull request by concretevitamin**
Sometimes we see this error in GCP provisioning (see e.g., https://github.com/skypilot-org/skypilot/issues/1006). Here we make sure the provisioning loop doesn't get crashed.

This seems hard to reproduce. Running
```
sky launch --use-spot --gpus A100:8 --down --cloud gcp -c dbg --region us-central1 -ry
```

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -655,6 +655,15 @@ def _update_blocklist_on_gcp_error(
             try:
                 exception_dict = ast.literal_eval(exception_str)
             except Exception as e:
+                if 'wait_ready timeout exceeded' in exception_str:
+                    # It's not clear we should block the current
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:659]:
```suggestion
                    # It's not clear if we should block the current
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for identifying the problem @concretevitamin! I am thinking we should retry the `ray_up` if the error message occur by adding it to the following place
https://github.com/skypilot-org/skypilot/blob/ca9fed18edcc4b3006c4e181f931862c97ef9c12/sky/backends/cloud_vm_ray_backend.py#L1686-L1704

With that retry logic added, we can probably block the zone in the `_update_blocklist_on_gcp_error`. Wdyt?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -3406,9 +3413,12 @@ def teardown_no_lock(self,
...
                                  f'"(labels.ray-cluster-name={cluster_name})" '
                                  f'--zones={zone} --format=value\\(name\\)')
+                    # If there are no instances, exit with 0 rather than causing
+                    # the delete command to fail.
                     terminate_cmd = (
-                        f'gcloud compute instances delete --zone={zone}'
-                        f' --quiet $({query_cmd})')
+                        f'NAMES=$({query_cmd}) && [ -n "$NAMES" ] && '
+                        f'gcloud compute instances delete --zone={zone} --quiet'
+                        ' $NAMES || echo "No instances to delete."')
```
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:3421]: Nice! Can we also make the AWS termination command robust, and set the L3388 to be `prev_cluster_status != status_lib.ClusterStatus.UP` to fix #2089. We can do it in another PR as well.

**Comment by concretevitamin** [comment]
> I am thinking we should retry the `ray_up` if the error message occur by adding it to the following place

Good point. One consideration is that, IIRC this wait_ready timeout can take ~15-20 minutes to happen, and it could mean (guessing) either (1) the zone is congested; (2) for AWS, ssh_proxy_command is set up incorrectly. If we retry ray_up() we may incur 15-20 minutes a few times before failing over. This is undesirable for both (1) and (2). Wdyt?

**Comment by Michaelvll** [comment]
> Good point. One consideration is that, IIRC this wait_ready timeout can take ~15-20 minutes to happen, and it could mean (guessing) either (1) the zone is congested; (2) for AWS, ssh_proxy_command is set up incorrectly. If we retry ray_up() we may incur 15-20 minutes a few times before failing over. This is undesirable for both (1) and (2). Wdyt?

That makes sense! If that is the case, should we add the zone to the blocklist then? Otherwise, we will go to the same zone again immediately due to the re-optimization and failover.

File: sky/backends/cloud_vm_ray_backend.py (hunk 2 above)
Comment (concretevitamin) [reply sky/backends/cloud_vm_ray_backend.py]: I added the empty-list handling to the AWS CLI invocation. Will leave #2089 to later which seems to require more involved handling (down the head node first).

**Comment by concretevitamin** [comment]
Addressed. PTAL.

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @concretevitamin! LGTM.

**Comment by concretevitamin** [comment]
Tested:
- [x] `pytest tests/test_smoke.py`
